woocommerce/woocommerce
# Add WooCommerce Job Queue using Action Scheduler for Async Webhook Delivery

This is a prototype to demonstrate what is needed to use Action Scheduler as the event queue for async webhook delivery (#18003), after [Slack discussion on the topic](https://woocommercecommunity.slack.com/archives/C4TNYTR28/p1525847140000049). I thought a PR would be a better channel for discussing and demonstrating it than Slack.

~The actual code is not intended to be merged as is, as the version of Action Scheduler included is a pre-release/dev version. If the use of Action Scheduler is considered to be a suitable solution to #18003, I will submit a different PR with the finalised version of Action Scheduler v2.0.~ *Edit:* PR has been updated.

### Commits of Note

Most of the 11,000 additions here can be ignored, because they are just adding the Action Scheduler library. The commits that matter are:

* SHA: 9101d02c2c4d4e - adds an async webhook delivery to the Action Scheduler queue
* SHA: 3531f26a17 - uses an Action Scheduler feature that isn't available with WP Cron or some other background processes to _group_ webhook deliveries, which allows them to be managed specifically rather than being batched with all other scheduled actions. This is handy on large sites that may want to run a specific webhook queue, and allow other actions to run in other queues.

### Outstanding Question

One difference between Action Scheduler's `wc_schedule_single_action()` and WordPress' `wp_schedule_single_event()` functions is that [with the later](https://developer.wordpress.org/reference/functions/wp_schedule_single_event/#description):

> scheduling an event to occur within 10 minutes of an existing event with the same action hook will be ignored unless you pass unique $args values for each scheduled event.

I believe that means that currently, webhooks scheduled for the same objects for different events that happen within 10 minutes (or more specifically, within the time between when the event occurs, and WP Cron processes the webhook) will not trigger two webhooks. Instead, only one will be triggered.

However, using `wc_schedule_single_action()`, both will be triggered (and the existing hook would need to be manually removed if desired).

For example, if an order's status is changed to on-hold, then processing within a few seconds, and WP Cron has not processed the webhook for the on-hold status change yet, then only the webhook for the processing status change would be triggered when using `wp_schedule_single_event()`. When using `wc_schedule_single_event()`, a webhook for both status transitions would be sent. This is very closely related to https://github.com/woocommerce/woocommerce/issues/15699, and using Action Scheduler with more data in the webhook's `$arg` may provide a solution for it.

### Manual Testing

A good way to test this is to run a process that creates a few thousand webhooks (or even better, 10,000s of webhooks), then watch Action Scheduler process them.

There is an admin UI to show the list and current status of actions, see https://github.com/Prospress/action-scheduler#managing-scheduled-actions

Running them via WP CLI would also be a good idea as a test site likely has WP Cron triggered infrequently, and out-of-the-box, Action Scheduler relies on WP Cron to process the queue. Docs on running a Action Scheduler queue via WP CLI here: https://github.com/Prospress/action-scheduler#wp-cli
